### PR TITLE
Add UniqueIds to generate UUIDs with fewer scalability bottlenecks

### PR DESCRIPTION
**Pull request by schlosna**
## General
**Before this PR**:

UUID generation and string serialization can be a scalability bottleneck, especially under high volume AtlasDB workloads. The top two I have seen recently pop up in profiles:

- `com.palantir.atlasdb.transaction.impl.TransactionRetryStrategy.runWithRetry(IntPredicate, TransactionRetryStrategy$Retryable):71`
  - https://github.com/palantir/atlasdb/blob/1acfd3d41b3e7e504320ea76b40ee9c134dab91a/atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategy.java#L71
- `com.palantir.lock.client.LeasedLockToken.of(ConjureLockToken, Lease):39`
  - https://github.com/palantir/atlasdb/blob/1acfd3d41b3e7e504320ea76b40ee9c134dab91a/lock-api/src/main/java/com/palantir/lock/client/LeasedLockToken.java#L39

image

There is some OpenJDK work either completed or in flight that is worth tracking:

- https://github.com/openjdk/jdk/pull/14135 (not yet integrated, closed as stale)
- https://github.com/openjdk/jdk/pull/14745
- https://github.com/openjdk/jdk/pull/14578

Also worth tracking and possibly utilizing one of the newer proposed UUID formats:

- https://uuid6.github.io/uuid6-ietf-draft/
- https://datatracker.ietf.org/doc/draft-peabody-dispatch-new-uuid-format/
- https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html#v8

**After this PR**:
==COMMIT_MSG==
Use tritium-ids UniqueIds to generate UUIDs with fewer scalability bottlenecks
==COMMIT_MSG==

```
16x Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz cores
VM version: JDK 17.0.8, OpenJDK 64-Bit Server VM, 17.0.8+7-LTS
Benchmark                             Mode  Cnt    Score     Error   Units
UuidBenchmark.jdkRandomUuidMax       thrpt    3    1.673 ?   1.378  ops/us // negative scale!
UuidBenchmark.jdkRandomUuidOne       thrpt    3    2.511 ?   0.272  ops/us
UuidBenchmark.jugSecureRandomMax     thrpt    3    2.008 ?   1.307  ops/us
UuidBenchmark.pseudoRandomUuidV4Max  thrpt    3  236.455 ? 106.688  ops/us
UuidBenchmark.pseudoRandomUuidV4One  thrpt    3   29.096 ?   3.265  ops/us
UuidBenchmark.randomUuidV4Max        thrpt    3   28.202 ?   8.763  ops/us
UuidBenchmark.randomUuidV4One        thrpt    3    3.388 ?   0.266  ops/us
UuidBenchmark.secureRandomMax        thrpt    3    2.015 ?   0.923  ops/us
UuidBenchmark.threadLocalMax         thrpt    3    1.652 ?   1.700  ops/us
UuidBenchmark.threadLocalSha1Max     thrpt    3   27.365 ?   5.918  ops/us
UuidBenchmark.twoLongsMax            thrpt    3    0.676 ?   0.450  ops/us
```

```
MacBook Pro 2021 10x aarch64 cores
VM version: JDK 17.0.8.1, OpenJDK 64-Bit Server VM, 17.0.8.1+8-LTS
Benchmark                             Mode  Cnt    Score    Error   Units
UuidBenchmark.jdkRandomUuidMax       thrpt    3    2.263 ?  0.358  ops/us // negative scale!
UuidBenchmark.jdkRandomUuidOne       thrpt    3    4.005 ?  0.211  ops/us
UuidBenchmark.jugSecureRandomMax     thrpt    3    2.902 ?  0.156  ops/us
UuidBenchmark.pseudoRandomUuidV4Max  thrpt    3  284.048 ? 48.930  ops/us
UuidBenchmark.pseudoRandomUuidV4One  thrpt    3   38.915 ? 52.533  ops/us
UuidBenchmark.randomUuidV4Max        thrpt    3   31.332 ?  4.457  ops/us
UuidBenchmark.randomUuidV4One        thrpt    3    4.583 ?  0.705  ops/us
UuidBenchmark.secureRandomMax        thrpt    3    2.934 ?  0.115  ops/us
UuidBenchmark.threadLocalMax         thrpt    3    2.296 ?  0.223  ops/us
UuidBenchmark.threadLocalSha1Max     thrpt    3   26.574 ? 12.350  ops/us
UuidBenchmark.twoLongsMax            thrpt    3    1.122 ?  0.405  ops/us
```

**Priority**: P1

**Concerns / possible downsides (what feedback would you like?)**:
Are all of the places using `pseudoRandomUuid()` safe or should they use `randomUuidV4`?

**Is documentation needed?**:

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.logsafe.Preconditions;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+
+public final class UniqueIds {
+    private UniqueIds() {}
+
+    public static UUID uuid() {
```
Comment (schlosna) [inline atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java]: - [ ] Add JavaDoc about tradeoffs as this uses `ThreadLocalRandom` so shouldn't be used where we need cryptographic random UUID

File: atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+package com.palantir.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.logsafe.Preconditions;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+
+public final class UniqueIds {
```
Comment (schlosna) [inline atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java]: - [x] Add JMH benchmark

Comment (schlosna) [reply atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java]: Need to benchmark against `java-uuid-generator` in #6725

File: atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java (hunk 1 above)
Comment (schlosna) [reply atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java]:
```suggestion
    public static UUID pseudoRandomUuid() {
```

File: atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    private UniqueIds() {}
+
+    public static UUID uuid() {
+        return uuid(ThreadLocalRandom.current());
+    }
+
```
Comment (schlosna) [inline atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java]: - [ ] Add v6, v7, v8 per https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html

**Comment by schlosna** [comment]
Pending https://github.com/palantir/tritium/pull/1789

File: atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java (hunk 2 above)
Comment (schlosna) [reply atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java]: Superseded by https://github.com/palantir/tritium/pull/1789

File: lock-api-objects/src/main/java/com/palantir/lock/v2/StartTransactionRequestV4.java [hunk 4]
```diff
@@ -33,7 +34,7 @@ public interface StartTransactionRequestV4 {
 
     static StartTransactionRequestV4 createForRequestor(UUID requestorUuid, int numTransactions) {
         return ImmutableStartTransactionRequestV4.builder()
-                .requestId(UUID.randomUUID())
+                .requestId(UniqueIds.randomUuidV4())
```
Comment (jeremyk-91) [inline lock-api-objects/src/main/java/com/palantir/lock/v2/StartTransactionRequestV4.java]: I think pseudorandom should do? This is the ID for the lease token we use to lock the immutable timestamp, so it should be the same as those

File: lock-api/src/main/java/com/palantir/lock/client/metrics/TimeLockFeedbackBackgroundTask.java [hunk 5]
```diff
@@ -51,7 +52,7 @@ public final class TimeLockFeedbackBackgroundTask implements AutoCloseable {
     private static final ScheduledExecutorService executor =
             PTExecutors.newSingleThreadScheduledExecutor(new NamedThreadFactory(TIMELOCK_FEEDBACK_THREAD_PREFIX, true));
 
-    private final UUID nodeId = UUID.randomUUID();
+    private final UUID nodeId = UniqueIds.randomUuidV4();
```
Comment (jeremyk-91) [inline lock-api/src/main/java/com/palantir/lock/client/metrics/TimeLockFeedbackBackgroundTask.java:55]: used by timelock for identifying clients as part of feedback - can be pseudorandom, though given it's only called once, doesn't really matter?

**Review by jeremyk-91 — APPROVED** [review]
👍 Looks great, thanks! There are probably at least two places where we didn't need to use the crypto secure one. For the start transactions we probably should change so that it's consistent with the locking APIs as that's what that lock request is used for; for the feedback service one, it's only generated once so no strong opinions

File: atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java (hunk 3 above)
Comment (schlosna) [reply atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java]: deferred to future once spec is final

File: lock-api/src/main/java/com/palantir/lock/client/metrics/TimeLockFeedbackBackgroundTask.java (hunk 5 above)
Comment (schlosna) [reply lock-api/src/main/java/com/palantir/lock/client/metrics/TimeLockFeedbackBackgroundTask.java:55]: yeah, I figured this wasn't on hot path so shouldn't matter.

File: atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java (hunk 1 above)
Comment (schlosna) [reply atlasdb-commons/src/main/java/com/palantir/util/UniqueIds.java]: OBE

File: lock-api-objects/src/main/java/com/palantir/lock/v2/StartTransactionRequestV4.java (hunk 4 above)
Comment (schlosna) [reply lock-api-objects/src/main/java/com/palantir/lock/v2/StartTransactionRequestV4.java]: thanks, updated

**Pull request merged**